### PR TITLE
Client power user API

### LINKED ISSUE #191: Client-side power API
The current generated client has an API suitable for message streams, but does not provide access to underlying protocol details.

For some use cases, such as adding per-call authentication principal information or reading custom response headers, we should have a 'power API'.  Likely we'll have a method on the 'normal' client to get an instance of the 'powerful' client, which would also have methods for each call, but those would then return an object on which further specifics can be specified before the call is issued, something like:

```
val myServiceClient: MyServiceClient
val s: Source[Foo] = myServiceClient.getFoos(query)
val a: AdvancedMyServiceClient = myServiceClient.advanced
val (s: Source[Foo], rh: Seq[HttpHeader]) = a.getFoos.addRequestHeader("Authentication", "Complicated").exposeResponseHeaders.execute()
```

(all naming and other specifics are still up for debate ;) )

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #191

Went for 
```
client.method()
  .bladi()
  .dadi()
  .execute(param)
```

instead of 
```
client.advanced.method(param)
  .bladi()
  .dadi()
  .execute()
```

### CODE REVIEW

**Comment by johanandren** [comment]
No usage samples yet, but the "normal" calls are now expressed through the new RequestBuilder version of the method, and interop-tests passing.

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ * Request builder for requests giving providing per call metadata capabilities
+ *
+ * Not for user extension
+ */
+@DoNotInherit
+sealed trait RequestBuilder[Req, Res] {
+
+  def addMetadata(key: String, value: String): RequestBuilder[Req, Res]
+  def withDeadline(deadline: FiniteDuration): RequestBuilder[Req, Res]
+  def execute(req: Req): Res
```
Comment (renatocaval) [inline runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: What about `invoke` instead of `execute`?

Comment (renatocaval) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Me neither. ;-) Lagom uses `invoke` hence my slight preference. 

Btw, that API style means that we always have a lifted API representation instead of a switch giving access to the lifted/advanced one. 
I thought we wanted to provide a non-lift, straight, variation for the most common cases.

**Comment by raboof** [comment]
Right! Moving the parameter to the end of the chain makes a lot of sense: this way a library user can even reuse the `RequestBuilder` for multiple requests that have the same characteristics. That seems like a cool advantage and I don't see any disadvantage.

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala (hunk 1 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Yeah. Not sure that it is always available rather than hidden behind a switch is a problem though. The old non-lifted API is still there (and should be the default)

Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Aligning with Lagom seems like sensible motivation for `invoke` :)

Comment (renatocaval) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Ah ok, now I see it. Because in gRPC there is always one param, we can overload the method with a paramless variation that gives access to the lifted version. Neat! Much better than a switch.

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt [hunk 2]
```diff
@@ -25,56 +26,30 @@ final class @{service.name}Client(settings: GrpcClientSettings)(implicit mat: Ma
...
-        ClientCalls.asyncBidiStreamingCall(
-          channel.newCall(@{method.name}Descriptor, options),
-          responseObserver)
-      }
-    in.via(flow)
+    }
   }
+
+  /**
+   * For access to method metadata use the parameter less version of ${method.name}
```
Comment (raboof) [inline codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: ('parameterless')

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  options: CallOptions,
+  delegatedExecute: (Req, CallOptions) => Ret) extends RequestBuilder[Req, Ret]() {
+
+  def addMetadata(key: String, value: String): RequestBuilderImpl[Req, Res, Ret] = {
+    // FIXME support values other than string values?
+    copy(options = options.withOption(CallOptions.Key.of(key, ""), value))
+  }
+
+  def withDeadline(deadline: FiniteDuration): RequestBuilderImpl[Req, Res, Ret] = {
+    copy(options = options.withDeadline(Deadline.after(deadline.length, deadline.unit)))
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Nice, we don't support this at the server side yet (#188) but since this is :+1:

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+/**
+ * INTERNAL API
+ */
+@InternalApi
+final case class RequestBuilderImpl[Req, Res, Ret](
+  options: CallOptions,
+  delegatedExecute: (Req, CallOptions) => Ret) extends RequestBuilder[Req, Ret]() {
+
+  def addMetadata(key: String, value: String): RequestBuilderImpl[Req, Res, Ret] = {
+    // FIXME support values other than string values?
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: do 'options' correspond to 'headers' or is there more indirection there?

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt [hunk 5]
```diff
@@ -25,56 +26,30 @@ final class @{service.name}Client(settings: GrpcClientSettings)(implicit mat: Ma
...
-        ClientCalls.asyncClientStreamingCall(
-          channel.newCall(@{method.name}Descriptor, options),
-          responseObserver)
+  /**
+   * Lower level version of the method, prefer @{method.name}(@method.parameterType) if possible.
+   */
+  def @{method.name}(): RequestBuilder[@method.parameterType, @method.returnType] = {
+    val descriptor = @{method.name}Descriptor
+    @if(method.methodType == akka.grpc.gen.Unary) {
+      RequestBuilderImpl.unary(descriptor, channel, options)
```
Comment (raboof) [inline codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: Do we want to organize the template so that the indentation of the generated code makes sense? No real preference here.

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala (hunk 4 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: I think there is a bit more indirection, there is also some marshalling infra in the grpc library around this. Not sure if we care about that though, maybe strings are good enough for now, and also avoids tying us down to something client specific. 

It's annotated as an experimental API, but so is most of these CallOptions.

File: codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt (hunk 5 above)
Comment (johanandren) [reply codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt]: Ah, I thought I did follow that, since it is how it was, but maybe having the templates readable is more important for maintainability than not having too many indentation levels in the generated sources?

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala (hunk 3 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Yeah, just picked this randomly to have some functionality to provide through the request builder.

**Comment by johanandren** [comment]
One thing I couldn't figure out was a reasonable way to deal with response metadata, for single response calls it may be relatively straight forward API wise (`Future[(Resp, Metadata)]`) but for streaming data it just doesn't make sense to put it inside of the stream, as you'd potentially want to act on it before consuming the response stream.

Any thoughts around that?

**Comment by ktoso** [comment]
I had proposed a wrapper type for that I believe; `GrpcResponse(stream or future).with(all the stuff)`, WDYT?

**Comment by johanandren** [comment]
Thinking about this some more, maybe the primary use case will be to _set_ metadata on the request, not deal with response metadata, so perhaps there should be the current "simple" `invoke(req)` and then a more complete but also more messy `invokeWithAnnoyingReturnType(req)`?

**Comment by renatocaval** [comment]
I think a better way of doing that is to have a method that changes the return type. That's also how it was done in Lagom.

For instance:
```
val res: Future[Res] =
  client.method()
    .addMetadata(key, value)
    .invoke(param)

val res: Future[GrpcResponse[Res]] =
  client.method()
    .addMetadata(key, value)
    .withResponseHeaders
    .invoke(param)
```

This requires two impl of `RequestBuilder` and the method `withResponseHeaders` works as a one direction switch.

**Comment by johanandren** [comment]
I don't see why a stateful/extra type is better than an additional `invoke` method with different return type. Am I missing something? Just to align with Lagom?

**Comment by TimMoore** [comment]
Yeah, I think it is pretty useful to be able to write decorators that can manipulate either/both the request and response metadata and payloads. I don't see a way with this API to be able to write a decorating function that can intercept the response without using internal APIs.

**Comment by johanandren** [comment]
Hmm, I still don't get it, why would a special type with transformations be better than just having a `Future[ResponseAndHeadersType]`/`CompletionStage<ResponseAndHeadersType>` which you can transform in exactly the same way as all your other async thingies and that you already know the API of?

**Comment by johanandren** [comment]
I think all those cases you describe should be covered by having two methods on the builder: the one that currently exists and then one that returns a `Future[ResponseWithHeaders]`. Just calling `map` on that then allows you to transform the response to any type you want while also looking at the response headers.

**Comment by renatocaval** [comment]
Sure, but does that mean that in case of streaming we will get a `Future[StreamResponseWithHeaders[T, R]]` from each I could read the headers and the `Source`?

This sounds good and it's worth experiment. I has some implications though.

 If the general API returns `Source[T, R]`, the power user API will have to return `Future[StreamResponseWithHeaders[T, R]]`.

It also means that there is effectively two calls, one to get the `Future` and one when we effectively materialize the `Source`, but I guess we can't avoid it. Unless we return `Source[(Headers, T), R]` which has an obvious overhead penalty.

And then we may have a another problem, for the moment mostly theoretical. This two calls may create a situation where the metadata differs between them.

**Comment by renatocaval** [comment]
For completeness and clarity. 

The implications I mentioned above (two calls) also holds for the `handleResponseHeader` API. I didn't mean to say that using `Future[ResponseWithHeaders]` was forcing us into that situation.

**Comment by renatocaval** [comment]
Just a crazy thought. 

The only way to associate headers to a stream and make sure that they relate (make part of the same call) is `Source[(Headers, T), R]`. We also want to keep the semantics of `Source` and be able to re-run it (make the call again) whenever we want.

So, maybe...

```scala
sealed trait Headers
case object EmptyHeaders extends Headers
case class NonEmptyHeaders extends Headers
```
Than is up to the implementers to decide when to send the headers. On the first element, on each element, etc.

Didn't thought thoroughly on the implications of this idea. Maybe stupid, maybe crazy, maybe good. I let this go through your brains first. :-)

**Comment by raboof** [comment]
I'm not entirely convinced we want to keep the semantics of Source and be able to re-run it (make the call again) whenever we want: I don't think we do that for `Source`-backed entities in `HttpRequest`s either.

`Source[(Headers, T), R]` suggests each `T` may have headers, which is not the case (as it's the response that has (leading and trailing) headers, not each element in the response individually) - so from that perspective returning the future containing headers and source seems to make more sense.

**Comment by renatocaval** [comment]
> suggests each `T` may have headers
You are totally right on that. Not a good idea.

**Comment by johanandren** [comment]
Internally the headers will arrive when the flow is materialized, and then there is a `Future[Trailers]` since those won't arrive until the stream completes. Haven't thought through how to provide this to the user yet, but the underlying stage I just wrote can at least be materialized multiple times and each materialization leads to one call - one set of metadata.

**Comment by ignasi35** [comment]
If a `Source` can be materialised several times causing a new call each time, then could headers and trailers be the materialised value? (I don't even know if that makes sense)

`Source[ T, ( Future[Headers], Future[Trailers] ) ]`

**Comment by johanandren** [comment]
Yeah, likely. But I think a specific type capturing it would be better than a tuple. The alternative would be to side effect from `mapMatValue` to complete some promises/futures for then, that way it could be something like `StreamingGrpcResponse(stream, futureHeaders, futureTrailers)`.

**Comment by ignasi35** [comment]
I find counterintuitive that materializing the `Source` several times causes new calls without using `client.method()...invoke(param)`.

IIRC Lagom's client produces a `Source` that's only meant to be materialized once and new calls only happen on new `invoke`. Also, in the 4 possible combinations (unary-basic, streamed-basic, unary-powerUser, stream-powerUser) I think this would be the only combination where materializing twice creates two calls, right? Or is a `Source` obtained using the basic (non-powerUSer) API also behaving like this?

Maybe renaming akka-grpc's `invoke` to `callFactory` or something that's not `invoke` could help make it more obvious what's actually happening.

**Comment by johanandren** [comment]
Both of the source variation could be materialized many times. If that is an issue we can protect against that and fail. I don't have a gut feeling about it being surprising or not yet, I'll see when the actual user API is in place (which is what I am currently working on).

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/LiftedGreeterClient.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    implicit val sys = ActorSystem()
+    implicit val mat = ActorMaterializer()
+    implicit val ec = sys.dispatcher
+
+    val client = new GreeterServiceClient(GrpcClientSettings(
+      "127.0.0.1",
+      8080,
+      overrideAuthority = Some("foo.test.google.fr"),
+      None,
```
Comment (raboof) [inline plugin-tester-scala/src/main/scala/example/myapp/helloworld/LiftedGreeterClient.scala]: perhaps name this parameter as well?

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala [hunk 7]
```diff
@@ -56,7 +66,7 @@ object GreeterClient {
       val responseStream = client.itKeepsReplying(HelloRequest("Alice"))
       val done: Future[Done] =
         responseStream.runForeach(reply => println(s"got streaming reply: ${reply.message}"))
-      Await.ready(done, 1.minute)
+      Await.ready(done, 1.minute) // just to keep sample simple - dont do Await.ready in actual code
```
Comment (raboof) [inline plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterClient.scala]: Perhaps we should avoid it here as well then - but let's keep that for another PR

File: runtime/src/main/scala/akka/grpc/GrpcClientResponse.scala [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.Optional
+import java.util.concurrent.CompletionStage
+
+import akka.annotation.DoNotInherit
+import io.grpc.Metadata
+
+import scala.concurrent.Future
+
+// FIXME should we provide our own immutable/thread safe response metadata abstraction?
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/GrpcClientResponse.scala]: Indeed the upstream `Metadata` is not too nice, might be good to at least wrap it in some immutable interface?.

File: runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+ * INTERNAL API
+ */
+@InternalApi
+trait OptionsModifications[T <: OptionsModifications[T]] {
+  def options: CallOptions
+  def updated(options: CallOptions): T
+
+  def addMetadata(key: String, value: String): T = {
+    // FIXME Key is instance equal to allow for replacing of the same key but still also allowing multiple
+    // values with the same name-key, not sure if it is important we support that
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala]: Let's keep an eye out if we ever see that (multiple `role`s?).

If so another way to do it might be to have this method overwrite keys, and introduce a `def addMetadata(key: String, value: Seq[String])` for multi-valued metadata

File: runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+trait OptionsModifications[T <: OptionsModifications[T]] {
+  def options: CallOptions
+  def updated(options: CallOptions): T
+
+  def addMetadata(key: String, value: String): T = {
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala]: Since later the `GrpcResponseMetadata` contains headers, perhaps `addHeader` would make more sense here?

File: runtime/src/main/scala/akka/grpc/internal/UnaryCallAdapter.scala [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  def future: Future[Res] = responsePromise.future
+  def cs: CompletionStage[Res] = future.toJava
+}
+
+/**
+ * gRPC Netty based client listener transforming callbacks into a future response
+ *
+ * INTERNAL API
+ */
+// needs to be a separate class because of CompletionStage error handling not bubling
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/UnaryCallAdapter.scala]: bubbling ;)

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+   * Invoke the gRPC method with the additional metadata added
+   *
+   * FIXME for streaming response this doesn't really make sense, left it to keep parity with single response but maybe it should go
+   */
+  def invoke(request: Req): Source[Res, NotUsed]
+
+  /**
+   * Invoke the gRPC method with the additional metadata added and provide access to response metadata
+   */
+  def invokeWithMetadata(request: Req): Source[Res, Future[GrpcResponseMetadata]]
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Encoding the metadata in the materialized value looks good.

It seems a 'clever' solution which sometimes backfires, but in this case it looks like it works out elegantly.

I agree since the metadata is in the materialized value the need for a separate `invokeWithMetadata` kind of disappears... unless perhaps there's opportunity for more optimization in `invoke` later?

File: runtime/src/main/scala/akka/grpc/internal/AkkaGrpcGraphStage.scala [hunk 13]
```diff
@@ -4,48 +4,188 @@
...
+ *  Flow in       ------ I------>  |                | ------- O ------->  Flow out
+ *                                 |  Netty client  |
+ *  upstream pull <-- onReady ---  |      call      | <-- request(1) ---  downstream pull
+ *                                 ------------------
+ *
+ *
+ * @param streamingResponse Do we expect a stream of responses or does more than 1 response mean a faulty server?
+ */
+@InternalApi
+private final class AkkaNettyGrpcClientGraphStage[I, O](
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/internal/AkkaGrpcGraphStage.scala]: cool

File: runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala (hunk 12 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/scaladsl/RequestBuilder.scala]: Yeah, that may be a good argument for keeping the two separate methods

File: runtime/src/main/scala/akka/grpc/internal/UnaryCallAdapter.scala (hunk 11 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/internal/UnaryCallAdapter.scala]: Less doubling in my bubbling!

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/LiftedGreeterClient.scala (hunk 6 above)
Comment (johanandren) [reply plugin-tester-scala/src/main/scala/example/myapp/helloworld/LiftedGreeterClient.scala]: ups, keep missing out on the plugin-testers as they are not part of the aggregated build

File: runtime/src/main/scala/akka/grpc/GrpcClientResponse.scala (hunk 8 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/GrpcClientResponse.scala]: Done!

File: runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala (hunk 9 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala]: It turned out that what I thought was the same thing, the CallOptions.options and the metadata headers, was in fact not at all the same thing, so this is gone now.

File: runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala (hunk 10 above)
Comment (johanandren) [reply runtime/src/main/scala/akka/grpc/internal/RequestBuilderImpl.scala]: Yeap, I renamed when changing it into metadata. I realize I renamed it wrong though, changed it to `withHeader` as I thought it replaces, but I realize now that it should be additive and not replace, so `add` would be better.

**Comment by johanandren** [comment]
Should be ready for merge now! :)

**Comment by johanandren** [comment]
Ah, crap, I missed that it maybe failed validation, github refuses to show me what failed right now though.

**Comment by raboof** [comment]
No worries, I think I fixed it on master
